allennlp/modules/transformer/t5.py

```diff
@@ -773,187 +801,473 @@ def forward(
+            T5LayerFF,
+            params=Params(
+                {
+                    "ff_proj": {
+                        "type": config.feed_forward_proj,
+                        "hidden_size": config.d_model,
+                        "ff_size": config.d_ff,
+                        "dropout": config.dropout_rate,
+                    },
+                    "layer_norm": layer_norm_kwargs,
+                    "dropout": config.dropout_rate,
+                }
+            ),
+        )
+        return {
+            "encoder": Lazy(
+                T5EncoderStack.basic_encoder,
+                contructor_extras={
+                    "num_blocks": config.num_layers,
+                    "block_self_attention": Lazy(T5Attention, contructor_extras=attention_kwargs),
+                    "final_layer_norm": T5LayerNorm(**layer_norm_kwargs),
+                    "block_ff": block_ff,
+                    "dropout": config.dropout_rate,
+                },
+            ),
+            "decoder": Lazy(
+                T5DecoderStack.basic_decoder,
+                contructor_extras={
+                    "num_blocks": config.num_decoder_layers,
+                    "block_self_attention": Lazy(T5Attention, contructor_extras=attention_kwargs),
+                    "block_cross_attention": Lazy(T5Attention, contructor_extras=attention_kwargs),
+                    "final_layer_norm": T5LayerNorm(**layer_norm_kwargs),
+                    "block_ff": block_ff,
+                    "dropout": config.dropout_rate,
+                },
+            ),
+            "decoder_start_token_id": config.decoder_start_token_id,
+            "pad_token_id": config.pad_token_id,
+            "eos_token_id": config.eos_token_id,
+            "vocab_size": config.vocab_size,
+            "model_dim": config.d_model,
+        }
+
+    def _shift_right(self, input_ids, start_value: int):
         # shift inputs to the right
         shifted_input_ids = input_ids.new_zeros(input_ids.shape)
         shifted_input_ids[..., 1:] = input_ids[..., :-1].clone()
-        shifted_input_ids[..., 0] = self.decoder_start_token_id
-
-        # replace possible -100 values in labels by `pad_token_id`
-        shifted_input_ids.masked_fill_(shifted_input_ids == -100, self.pad_token_id)
-
-        assert torch.all(
-            shifted_input_ids >= 0
-        ).item(), "Verify that `shifted_input_ids` has only positive values"
+        shifted_input_ids[..., 0] = start_value
 
         return shifted_input_ids
 
+    def _get_lm_logits(self, decoder_last_hidden_state: FloatT) -> FloatT:
+        # Shape: (batch_size, target_length, model_dim)
+        sequence_output = decoder_last_hidden_state
+        # Rescale output before projecting on vocab
+        # TODO: HF only does this when does this when embeddings are tied.
+        # Currently tied embeddings is the only option we have, but if make
+        # that configurable then we should put this in an 'if' block.
```

Comment:
The T5 paper does a bunch of experiments tying things here and there, but ultimately concludes that almost none of them work. The HF code supports all the weight tying options, but I removed that stuff because I just wanted to run T5-Large at the time. I suppose it would be good to have for someone to do future experiments, but not very critical.